### ISSUE TITLE
Add support for Inventory Task generation

### ISSUE THREAD

**Issue by githengi** [OpenSRP/opensrp-plan-evaluator#83]
Add support for Inventory Task generation

- [ ] Add Inventory Subject codable concept

- [ ] Add interfaces methods for getting data for Inventory 

- [ ] Implement the interfaces in Server core

- [ ] Implement the interfaces in Client core

**Comment by rehammuzzamil** [comment]
@cafootitt  The issue's development is on hold, due to the finalization of the mapping of the Stock entity to the FIHR domain resource.

**Comment by cafootitt** [comment]
Reham is blocked on this until the tasking lib work is completed.

**Comment by githengi** [comment]
The mapping of Stock/Inventory to be to [FHIR Basic Resource](https://www.hl7.org/fhir/basic.html)

Use this [mapping](https://docs.google.com/spreadsheets/d/1bFef_DmWt9_SrwUBKHhkKFVdZ0X0qYKbS_Rg-GfCyik/edit#gid=0) for the conversion.

This will  allow task generation to be varied on product catalog, or the inventory itself

**Comment by cafootitt** [comment]
@craigappl is reviewing the FHIR inventory mapping before this issue can be unblocked.

**Comment by craigappl** [comment]
@githengi , I want to recommend a different route. I'm sorry for all of the back and fourth on this, but I think it's important to get this right. I think we can map everything to the ["device" resource](https://www.hl7.org/fhir/device.html). A device is any manufactured item that can be used in the provision of healthcare without being substantially changed throughout the activity. So, a bicycle would be categorized as a device, but a box of medium latext gloves would not. You could use the [DeviceDefinition resource](https://www.hl7.org/fhir/devicedefinition.html) to define the list of catalogue entries from UNICEF that we would expect and then reference that tin the device.defintion field. Note that the [device-kind value set](http://hl7.org/fhir/valueset-device-kind.html) includes thing that are consumable like gloves, so it looks like it would match our use case pretty well.

There's precedent outlined in this thread on Zulip: https://chat.fhir.org/#narrow/stream/179165-committers/topic/inventory.20tracking (create an account with you Ona Google account)

**Comment by githengi** [comment]
Thanks @craigappl could you help come with a procedure to map to the valueset, we would need to map the product catalog to the value set and also handle cases where the current value has no mapping

**Comment by craigappl** [comment]
@githengi , I hand wrote a FHIR resource. It wouldn't pass validation because of the new fields I added. This may be a good thing to reference.

```
{
  "resourceType": "Device",
  "id": "c009ded9-1517-4784-b0ba-534d2306eca9",
  "type": {
    "coding": [
      {
        "system": "http://snomed.info/sct",
        "code": "52291003",
        "display": "Gloves"
      }
    ],
    "text": "Gloves"
  },
  "owner": "organization/unicefSection",
  "location": "location/Location1234",
  "serialNumber": "HN3873682-92787",
  "status": "active",
  "quantity": 50,
  "occurrenceDateTime": "2020-10-07T13:28:17+03:00",
  "supplier": "organization/supplyingOrg123",
  "poNumber": [
    {
      "use": "usual",
      "type": {
        "coding": [
          {
            "system": "http://hl7.org/fhir/ValueSet/identifier-type",
            "code": "PLAC"
          }
        ]
      },
      "value": "12345"
    }
  ]
}
```

**Comment by githengi** [comment]
@craigappl the coding will be a problem as it forces us to define the code on the catalog before hand. Also this needs to pass validation to make it usable, We should avoid adding custom fields and use extensions as we would have to update the FHIR library

**Comment by rehammuzzamil** [comment]
Mapping is defined here: https://docs.google.com/spreadsheets/d/1bFef_DmWt9_SrwUBKHhkKFVdZ0X0qYKbS_Rg-GfCyik/edit#gid=0

**Comment by craigappl** [comment]
Sam G and I talked today, I will create a bundle and document it here in the ticket so we can get it ready for work on Monday.

**Comment by rehammuzzamil** [comment]
Hi @githengi , I have created a converter using the mapping provided earlier in this sheet: https://docs.google.com/spreadsheets/d/1bFef_DmWt9_SrwUBKHhkKFVdZ0X0qYKbS_Rg-GfCyik/edit#gid=0 

Do we need to change its mapping and map Stock to SuuplyDelivery?

**Comment by githengi** [comment]
yes

**Comment by cafootitt** [comment]
Hoping to wind up and submit PR by COB Wed - Reham.

**Comment by craigappl** [comment]
@rehammuzzamil and @githengi , here's the sample coding of a bundle that validates against HAPI FHIR:
I also updated the spreadsheet: https://docs.google.com/spreadsheets/d/1bFef_DmWt9_SrwUBKHhkKFVdZ0X0qYKbS_Rg-GfCyik/edit#gid=0&range=G1

```
{
  "resourceType": "Bundle",
  "type": "collection",
  "entry": [
    {
      "fullUrl": "https://fhir.smartregister.org/device/00868475000235-d9c070d1-9e8b-46e8-b24a-a98ccb69aadd",
      "resource": {
        "resourceType": "Device",
        "id": "00868475000235-d9c070d1-9e8b-46e8-b24a-a98ccb69aadd",
        "owner": {
          "reference": "Organization/91a718fb-c903-394a-9c7d-fdef49a3a817"
        },
        "location": {
          "reference": "Location/7fdd41eb-4876-41e9-9938-ee7b8245a3fc"
        },
        "serialNumber": "SN1234321",
        "status": "active"
      }
    },
    {
      "fullUrl": "https://fhir.smartregister.org/supplyDelivery/b93c856f-1bac-4cf8-ad98-a2e8096e9dbc",
      "resource": {
        "resourceType": "SupplyDelivery",
        "id": "b93c856f-1bac-4cf8-ad98-a2e8096e9dbc",
        "identifier": [
          {
            "value": "Purchase Order 987654321"
          }
        ],
        "status": "completed",
        "supplier": {
          "reference": "Organization/1704261",
          "display": "Supplier 1"
        },
        "type": {
          "coding": [
            {
              "system": "http://terminology.hl7.org/CodeSystem/supply-item-type",
              "code": "device"
            }
          ]
        },
        "suppliedItem": {
          "quantity": {
            "value": 50
          },
          "itemCodeableConcept": {
            "coding": [
              {
                "code": "Latex Gloves",
                "display": "Latex Gloves Medium"
              }
            ]
          }
        },
        "occurrenceDateTime": "2020-12-07T23:17:56.558+00:00"
      }
    }
  ]
}
```

**Comment by githengi** [comment]
The stockId should also be mapped to the bundle id @rehammuzzamil

**Comment by githengi** [comment]
@craigappl  using a bundle is causing compile issues as most resources being used extend Domain resource to store contained resources. 
I propose we use Supply delivery as the main resource and then Device to be the contained resources instead of making a lot of changes to accommodate bundle which extends Resource and not domain Resource

**Comment by rehammuzzamil** [comment]
I assume to map UNICEF Section and Donor value to the owner and supplier fields in Device and SupplyDelivery resources respectively. Since we do not have any identifier against the Unicef section and donor as both are added to the system in Setting Configuration via settings endpoint. Please correct me if there is anything wrong.
cc: @githengi

**Comment by rehammuzzamil** [comment]
You can find updated plandefinition here https://github.com/OpenSRP/web/issues/180#issuecomment-741764788
@bennsimon  @p-netm  @cafootitt

**Comment by rehammuzzamil** [comment]
I have identified a minor issue while testing in PlanDefinition, I have updated the plan definition [here](https://github.com/OpenSRP/web/issues/180#issuecomment-742539468).

**Comment by rehammuzzamil** [comment]
Please find updated Plan Definition here https://github.com/OpenSRP/web/issues/180#issuecomment-743172261

**Comment by rehammuzzamil** [comment]
_Implementation details:_
Based on the plan activation, tasks for Service Point Check, Record GPS, and Product Check would be generated. 
For the Fix Product Problem task, plan evaluation will be triggered on a `flag_problem` event submission, where an event for inventory items will have a baseEntityId pointing to stock Id (not the DB PK). The stockId would reside inside questionnaire response and can be read through the following expression:
`$this.subject.reference.value`
